File: packages/sdk/agentkit/runtime/providers/registry.py
import importlib
from typing import Dict, Type

from agentkit.runtime.providers.base import ModelProvider
from agentkit.runtime.providers.errors import ProviderConfigError

_PROVIDERS: Dict[str, Type[ModelProvider]] = {}
_BUILTINS_LOADED = False


def register_provider(name: str, provider_cls: Type[ModelProvider]) -> None:
    _PROVIDERS[name] = provider_cls


def _ensure_builtin_providers_loaded() -> None:
    global _BUILTINS_LOADED
    if _BUILTINS_LOADED:
        return

    # Provider modules self-register through register_provider(...)
    importlib.import_module("agentkit.runtime.providers.gemini")
    importlib.import_module("agentkit.runtime.providers.kimi")
    importlib.import_module("agentkit.runtime.providers.openai")
    importlib.import_module("agentkit.runtime.providers.mock")
    _BUILTINS_LOADED = True


def get_provider(name: str, **kwargs) -> ModelProvider:
    _ensure_builtin_providers_loaded()
    provider_cls = _PROVIDERS.get(name)
    if not provider_cls:
        raise ProviderConfigError(f"Provider '{name}' is not registered.")
    return provider_cls(**kwargs)

File: agentkit/ts/src/providers/registry.ts
/**
 * Provider registry: a process-level map from name to `ModelProvider`.
 * Mirrors `agentkit.runtime.providers.registry`.
 */
import type { ModelProvider } from "./base";

const providers = new Map<string, ModelProvider>();

/** Register a provider under a name. Throws on duplicate. */
export function registerProvider(name: string, provider: ModelProvider): void {
  if (providers.has(name)) {
    throw new Error(`Provider already registered: ${name}`);
  }
  providers.set(name, provider);
}

/** Retrieve a registered provider. Throws if not found. */
export function getProvider(name: string): ModelProvider {
  const p = providers.get(name);
  if (p === undefined) {
    throw new Error(
      `Unknown provider: ${name}. Register it with registerProvider() first.`,
    );
  }
  return p;
}

/** Clear all registrations. For tests. */
export function clearProviders(): void {
  providers.clear();
}
